core/include/detray/surface_finders/grid/grid.hpp

```diff
@@ -0,0 +1,309 @@
+                           detail::grid_view<bin_storage_type>>;
+
+    /// Make grid default constructible: Empty grid with empty axis
+    grid() = default;
+
+    /// Create empty grid with empty axes from specific vecmem memory resource
+    DETRAY_HOST
+    grid(vecmem::memory_resource &resource)
+        : m_data(resource), m_axes(resource) {}
+
+    /// Create grid with well defined @param axes and @param bins_data - move
+    DETRAY_HOST_DEVICE
+    grid(bin_storage_type &&bin_data, axes_type &&axes)
+        : m_data(std::move(bin_data)), m_axes(std::move(axes)) {}
+
+    /// Create grid from container pointers - non-owning (both grid and axes)
+    DETRAY_HOST_DEVICE
+    grid(bin_storage_type *bin_data_ptr, axes_type &axes,
+         const dindex offset = 0)
+        : m_data(bin_data_ptr, offset), m_axes(axes) {}
+
+    /// Device-side construction from a vecmem based view type
+    template <typename grid_view_t,
+              typename std::enable_if_t<
+                  std::is_base_of_v<dbase_view, grid_view_t>, bool> = true>
+    DETRAY_HOST_DEVICE grid(grid_view_t &view)
+        : m_data(detray::detail::get<0>(view.m_views)),
+          m_axes(detray::detail::get<1>(view.m_views)) {}
+
+    /// @returns the underlying bin value storage. Either the container
+    /// or a container pointer to a global collection - const
+    DETRAY_HOST_DEVICE
+    auto data() const -> const storage_type & { return m_data; }
+
+    /// @returns the underlying bin value storage. Either the container
+    /// or a container pointer to a global collection - non-const for vecmem
+    // TODO: Don't do
+    DETRAY_HOST_DEVICE
+    auto data() -> storage_type & { return m_data; }
+
+    /// @returns the multi-axis used by the grid - const
+    DETRAY_HOST_DEVICE
+    auto axes() const -> const axes_type & { return m_axes; }
+
+    /// @returns the multi-axis used by the grid - non-const for vecmem
+    // TODO: Don't do
+    DETRAY_HOST_DEVICE
+    auto axes() -> axes_type & { return m_axes; }
+
+    /// @returns an axis object, corresponding to the label.
+    template <axis::label L>
+    DETRAY_HOST_DEVICE inline constexpr auto get_axis() const {
+        return m_axes.template get_axis<L>();
+    }
+
+    /// @returns an axis object of the given type.
+    template <typename axis_t>
+    DETRAY_HOST_DEVICE inline constexpr axis_t get_axis() const {
+        return m_axes.template get_axis<axis_t>();
+    }
+
+    /// Transform a point in global cartesian coordinates to local coordinates
+    ///
+    /// @param trf the placement transform of the grid (e.g. from a volume or
+    ///            a surface).
+    /// @param p   the point in global coordinates
+    ///
+    /// @returns a point in the coordinate system that is spanned by the grid's
+    /// axes.
+    template <typename transform_t, typename point3_t>
+    auto global_to_local(const transform_t &trf, const point3_t &p) const {
```

Comment:
Will it be replaced by the function of coordinate classes?

Reply:
Yes, what is currently the ```local_frame``` should become a coordinate class, once they are in

Reply:
It is defined with the multi-axis though, since it is really the axes that span the local coordinates